src/fpm_compiler.F90

```diff
@@ -1104,6 +1107,18 @@ subroutine compile_fortran(self, input, output, args, log_file, stat)
         & echo=self%echo, verbose=self%verbose, redirect=log_file, exitstat=stat)
 end subroutine compile_fortran
 
+subroutine generate_shared_library(self, package_name, output_file)
+    !> Instance of the compiler object
+    class(compiler_t), intent(in) :: self
+    !> Name of the package.
+    character(len=*), intent(in) :: package_name
+    !> Output file of library archive.
+    character(len=*), intent(in) :: output_file
+
+    call run(self%fc // " --shared " // " -o " // "lib" // package_name // ".so" // " " &
```

Comment:
Would this command work for all compilers in the `fpm` suite, or there are compiler specific flags that should be used? 

Also the `.so` output looks a bit odd to me: isn't a dynamic library supposed to be named `.dll` on Windows and `.dylib` on macOS?

Reply:
[This thread](https://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-macos) on SO seems to indicate `.dylib` as the canonical "dynamic library", whereas ".so" is used for plugins. In practice the formats are the same. The [Mac Dynamic Library Usage Guidelines](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/DynamicLibraryUsageGuidelines.html) refer to `.dylib` in many places, but doesn't mention `.so`?

So I think `.dylib` and the flag `-dynamiclib` instead of `-shared` would be the preferred option for MacOS.